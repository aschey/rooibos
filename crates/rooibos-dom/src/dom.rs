use std::cell::{OnceCell, RefCell};
use std::fmt;
use std::rc::Rc;
use std::sync::atomic::{AtomicU32, Ordering};

use leptos_reactive::{create_render_effect, untrack_with_diagnostics};
use ratatui::prelude::{Constraint, Direction, Layout, Rect};
use ratatui::Frame;
use slotmap::{new_key_type, SlotMap};

thread_local! {
   pub static DOM: RefCell<DomNode> = RefCell::new(DomNode::root());
   static DOM_NODES: RefCell<SlotMap<DomNodeKey, DomNodeInner>> =
    RefCell::new(SlotMap::<DomNodeKey, DomNodeInner>::default());
}

static NODE_ID: AtomicU32 = AtomicU32::new(1);

pub trait IntoView {
    fn into_view(self) -> View;
}

pub trait Mountable {
    fn get_mountable_node(&self) -> DomNode;
}

#[derive(Clone, PartialEq, Eq)]
pub enum View {
    DynChild(DynChildRepr),
    Component(ComponentRepr),
    DomNode(DomNode),
    DomWidget(DomWidget),
}

impl fmt::Debug for View {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        todo!()
    }
}

#[derive(Clone, PartialEq, Eq)]
enum NodeType {
    Layout { direction: Direction, margin: u16 },
    Root,
    Overlay,
    Widget(DomWidget),
}

new_key_type! {struct DomNodeKey; }

impl IntoView for View {
    fn into_view(self) -> View {
        self
    }
}

impl Mountable for View {
    fn get_mountable_node(&self) -> DomNode {
        match self {
            Self::DomNode(dom_node) => dom_node.clone(),
            Self::DynChild(dyn_child) => dyn_child.get_mountable_node(),
            Self::Component(component) => component.get_mountable_node(),
            Self::DomWidget(widget) => widget.get_mountable_node(),
        }
    }
}

impl IntoView for &View {
    fn into_view(self) -> View {
        self.clone()
    }
}

impl<const N: usize> IntoView for [View; N] {
    fn into_view(self) -> View {
        Fragment::new(self.into_iter().collect()).into_view()
    }
}

pub struct Fragment {
    id: u32,
    /// The nodes contained in the fragment.
    pub nodes: Vec<View>,
    pub(crate) view_marker: Option<String>,
}

impl Fragment {
    pub fn new(nodes: Vec<View>) -> Self {
        Self {
            id: NODE_ID.fetch_add(1, Ordering::Relaxed),
            nodes,
            #[cfg(debug_assertions)]
            view_marker: None,
        }
    }
}

impl FromIterator<View> for Fragment {
    fn from_iter<T: IntoIterator<Item = View>>(iter: T) -> Self {
        Fragment::new(iter.into_iter().collect())
    }
}

impl From<View> for Fragment {
    fn from(view: View) -> Self {
        Fragment::new(vec![view])
    }
}

impl From<Fragment> for View {
    fn from(value: Fragment) -> Self {
        let mut frag = ComponentRepr::new_with_id("".to_string(), value.id);

        frag.view_marker = value.view_marker;
        frag.children = value.nodes;

        frag.into_view()
    }
}

impl IntoView for Fragment {
    fn into_view(self) -> View {
        self.into()
    }
}

#[derive(Clone, PartialEq, Eq)]
pub struct ComponentRepr {
    pub(crate) document_fragment: DocumentFragment,
    mounted: Rc<OnceCell<DomNode>>,
    pub(crate) name: String,
    _opening: Comment,
    pub children: Vec<View>,
    closing: Comment,
    pub(crate) id: u32,
    pub(crate) view_marker: Option<String>,
}

impl ComponentRepr {
    pub fn new_with_id(name: String, id: u32) -> Self {
        let document_fragment = DocumentFragment::widget(DomWidget::new(|_, _| {}));
        let markers = (
            Comment::new(format!("<{name}>"), id, false),
            Comment::new(format!("</{name}>"), id, true),
        );

        Self {
            document_fragment,
            mounted: Default::default(),
            _opening: markers.0,
            closing: markers.1,
            name,
            children: vec![],
            id,
            view_marker: None,
        }
    }
}

impl Mountable for ComponentRepr {
    fn get_mountable_node(&self) -> DomNode {
        if let Some(mounted) = self.mounted.get() {
            mounted.clone()
        } else {
            let node = self.document_fragment.clone().into_node();
            self.mounted.set(node.clone()).unwrap();
            node
        }
    }
}

impl IntoView for ComponentRepr {
    fn into_view(self) -> View {
        for child in &self.children {
            mount_child(MountKind::Before(&self.closing.node), child);
        }

        View::Component(self)
    }
}

pub enum MountKind<'a> {
    Before(
        // The closing node
        &'a DomNode,
    ),
    Append(&'a DomNode),
}

fn mount_child<M: Mountable + std::fmt::Debug>(kind: MountKind, child: &M) -> DomNodeKey {
    let child = child.get_mountable_node();

    match kind {
        MountKind::Append(el) => {
            el.append_child(&child);
        }
        MountKind::Before(closing) => {
            closing.before(&child);
        }
    }
    child.key
}

// #[derive(Clone, PartialEq, Eq)]
// pub struct DomNode {
//     id: Option<String>,
//     node_type: NodeType,
//     constraint: Constraint,
//     children: Vec<DomNode>,
// }

pub struct Component<F, V>
where
    F: FnOnce() -> V,
    V: IntoView,
{
    id: u32,
    name: String,
    children_fn: F,
}

impl<F, V> Component<F, V>
where
    F: FnOnce() -> V,
    V: IntoView,
{
    /// Creates a new component.
    pub fn new(name: impl Into<String>, f: F) -> Self {
        Self {
            id: NODE_ID.fetch_add(1, Ordering::Relaxed),
            name: name.into(),
            children_fn: f,
        }
    }
}

impl<F, V> IntoView for Component<F, V>
where
    F: FnOnce() -> V,
    V: IntoView,
{
    fn into_view(self) -> View {
        let Self {
            id,
            name,
            children_fn,
        } = self;

        let mut repr = ComponentRepr::new_with_id(name, id);

        // disposed automatically when the parent scope is disposed
        let child = untrack_with_diagnostics(|| children_fn().into_view());

        repr.children.push(child);

        repr.into_view()
    }
}

// #[derive(Clone)]
// pub struct View {
//     id: u32,
//     f: Rc<RefCell<dyn FnMut(&mut Frame, Rect)>>,
// }

pub struct DynChild<CF, V>
where
    CF: Fn() -> V + 'static,
    V: IntoView,
{
    id: u32,
    child_fn: CF,
}

impl<CF, N> DynChild<CF, N>
where
    CF: Fn() -> N + 'static,
    N: IntoView,
{
    pub fn new(child_fn: CF) -> Self {
        Self {
            child_fn,
            id: NODE_ID.fetch_add(1, Ordering::Relaxed),
        }
    }
}

impl<CF, N> IntoView for DynChild<CF, N>
where
    CF: Fn() -> N + 'static,
    N: IntoView,
{
    fn into_view(self) -> View {
        fn create_dyn_view(
            component: DynChildRepr,
            child_fn: Box<dyn Fn() -> View>,
        ) -> DynChildRepr {
            let closing = component.closing.node.clone();
            let child = component.child.clone();

            create_render_effect(move |prev_run: Option<Option<DomNodeKey>>| {
                let new_child = child_fn().into_view();
                let mut child_borrow = (*child).borrow_mut();

                // Is this at least the second time we are loading a child?
                if let Some(prev_t) = prev_run {
                    let child = child_borrow.take().unwrap();

                    if child != new_child {
                        let new = mount_child(MountKind::Before(&closing), &new_child);

                        **child_borrow = Some(new_child);
                        Some(new)
                    } else {
                        prev_t
                    }
                } else {
                    let new = mount_child(MountKind::Before(&closing), &new_child);
                    Some(new)
                }
            });
            component
        }

        let Self { id, child_fn } = self;

        let component = DynChildRepr::new_with_id(id);
        let component = create_dyn_view(component, Box::new(move || child_fn().into_view()));

        View::DynChild(component)
    }
    // fn into_view(self) -> View {
    //     let child_fn = self.child_fn;
    //     let prev = leptos_reactive::SpecialNonReactiveZone::enter();
    //     let view = child_fn().into_view();
    //     leptos_reactive::SpecialNonReactiveZone::exit(prev);

    //     create_render_effect(move |prev_id: Option<u32>| {
    //         let new_view = child_fn().into_view();
    //         let new_id = new_view.id;
    //         if let Some(prev_id) = prev_id {
    //             // if prev_id != new_id {
    //             DOM.with(|d| {
    //                 let res = d.borrow().replace(prev_id, new_view);
    //                 // dbg!(res.is_none());
    //             })
    //             // }
    //         }

    //         new_id
    //     });
    // }
}

#[derive(Clone, PartialEq, Eq)]
struct Comment {
    node: DomNode,
    content: String,
}

impl Comment {
    fn new(content: impl Into<String>, id: u32, closing: bool) -> Self {
        let node = DomNode::from_fragment(DocumentFragment::widget(DomWidget::new(|_, _| {})));
        DOM.with(|d| d.borrow().append_child(&node));
        Self {
            content: content.into(),
            node,
        }
    }
}

#[derive(Clone, PartialEq, Eq)]
pub struct DynChildRepr {
    document_fragment: DocumentFragment,
    #[cfg(debug_assertions)]
    opening: Comment,
    pub(crate) child: Rc<RefCell<Box<Option<View>>>>,
    closing: Comment,
    pub(crate) id: u32,
}

impl DynChildRepr {
    fn new_with_id(id: u32) -> Self {
        let document_fragment = DocumentFragment::widget(DomWidget::new(|_, _| {}));
        let markers = (
            Comment::new("<DynChild>", id, false),
            Comment::new("</DynChild>", id, true),
        );

        Self {
            document_fragment,
            opening: markers.0,
            closing: markers.1,
            child: Default::default(),
            id,
        }
    }
}

impl Mountable for DynChildRepr {
    fn get_mountable_node(&self) -> DomNode {
        self.document_fragment.clone().into_node()
    }
}

// impl Eq for View {}

// impl PartialEq for View {
//     fn eq(&self, other: &Self) -> bool {
//         self.id == other.id
//     }
// }

// impl<F, V> IntoView for F
// where
//     F: Fn() -> V + 'static,
//     V: IntoView,
// {
//     fn into_view(self) -> View {
//         DynChild::new(self).into_view()
//     }
// }

// impl View {
//     pub fn new(f: impl FnMut(&mut Frame, Rect) + 'static) -> Self {
//         Self {
//             id: 0,
//             f: Rc::new(RefCell::new(f)),
//         }
//     }

//     pub fn render(&self, frame: &mut Frame, rect: Rect) {
//         (self.f.borrow_mut())(frame, rect)
//     }
// }

// impl<F: 'static> IntoView for F
// where
//     F: FnMut(&mut Frame, Rect),
// {
//     fn into_view(self) -> View {
//         View::new(self)
//     }
// }

#[derive(Clone)]
pub struct DomWidget {
    f: Rc<RefCell<dyn FnMut(&mut Frame, Rect)>>,
}

impl DomWidget {
    pub fn new<F: FnMut(&mut Frame, Rect) + 'static>(f: F) -> Self {
        Self {
            f: Rc::new(RefCell::new(f)),
        }
    }

    pub fn render(&self, frame: &mut Frame, rect: Rect) {
        (*self.f).borrow_mut()(frame, rect)
    }
}

impl IntoView for DomWidget {
    fn into_view(self) -> View {
        View::DomWidget(self)
    }
}

impl Mountable for DomWidget {
    fn get_mountable_node(&self) -> DomNode {
        DomNode::from_fragment(DocumentFragment::widget(self.clone()))
    }
}

impl PartialEq for DomWidget {
    fn eq(&self, other: &Self) -> bool {
        false
    }
}

impl Eq for DomWidget {}

impl<F, N> IntoView for F
where
    F: Fn() -> N + 'static,
    N: IntoView,
{
    fn into_view(self) -> View {
        DynChild::new(self).into_view()
    }
}

#[derive(Clone, PartialEq, Eq)]
pub struct DocumentFragment {
    node_type: NodeType,
    constraint: Constraint,
    children: Vec<DomNodeKey>,
}

impl DocumentFragment {
    pub fn widget(widget: DomWidget) -> Self {
        Self {
            node_type: NodeType::Widget(widget),
            children: vec![],
            constraint: Constraint::Min(0),
        }
    }

    pub fn row() -> Self {
        Self {
            node_type: NodeType::Layout {
                direction: Direction::Horizontal,
                margin: 0,
            },
            children: vec![],
            constraint: Constraint::Min(0),
        }
    }

    pub fn col() -> Self {
        Self {
            node_type: NodeType::Layout {
                direction: Direction::Vertical,
                margin: 0,
            },
            children: vec![],
            constraint: Constraint::Min(0),
        }
    }

    pub fn overlay() -> Self {
        Self {
            node_type: NodeType::Overlay,
            children: vec![],
            constraint: Constraint::Min(0),
        }
    }

    pub fn constraint(mut self, constraint: Constraint) -> Self {
        self.constraint = constraint;
        self
    }

    pub fn margin(mut self, new_margin: u16) -> Self {
        if let NodeType::Layout { margin, .. } = &mut self.node_type {
            *margin = new_margin;
        }
        self
    }

    pub fn into_node(self) -> DomNode {
        DomNode::from_fragment(self)
    }
}

pub struct Element {
    inner: DomNode,
}

impl Element {
    pub fn child(self, child: impl IntoView) -> Self {
        let child = child.into_view();
        mount_child(MountKind::Append(&self.inner), &child);
        self
    }

    pub fn constraint(self, constraint: Constraint) -> Self {
        DOM_NODES.with(|d| d.borrow_mut()[self.inner.key].constraint = constraint);
        self
    }
}

impl IntoView for Element {
    fn into_view(self) -> View {
        View::DomNode(self.inner)
    }
}

pub fn row() -> Element {
    Element {
        inner: DocumentFragment::row().into_node(),
    }
}

pub fn col() -> Element {
    Element {
        inner: DocumentFragment::col().into_node(),
    }
}

#[derive(Clone, PartialEq, Eq)]
struct DomNodeInner {
    node_type: NodeType,
    constraint: Constraint,
    children: Vec<DomNodeKey>,
    parent: Option<DomNodeKey>,
}

impl DomNodeInner {
    pub fn render(&self, frame: &mut Frame, rect: Rect) {
        DOM_NODES.with(|d| {
            let d = d.borrow();
            let children: Vec<_> = self.children.iter().map(|c| &d[*c]).collect();
            match &self.node_type {
                NodeType::Layout { direction, margin } => {
                    let layout = Layout::default().direction(*direction).margin(*margin);
                    let constraints = children.iter().map(|c| c.constraint);
                    let chunks = layout
                        .constraints(constraints.collect::<Vec<_>>())
                        .split(rect);
                    children
                        .iter()
                        .zip(chunks.iter())
                        .for_each(|(child, chunk)| {
                            child.render(frame, *chunk);
                        });
                }
                NodeType::Overlay | NodeType::Root => {
                    children.iter().for_each(|child| {
                        child.render(frame, rect);
                    });
                }
                NodeType::Widget(widget) => {
                    widget.render(frame, rect);
                }
            }
        });
    }
}

#[derive(Clone, PartialEq, Eq, Debug)]
pub struct DomNode {
    // inner: Rc<RefCell<DomNodeInner>>,
    key: DomNodeKey,
    // parent: Rc<RefCell<Option<DomNodeKey>>>,
}

impl DomNode {
    pub fn root() -> Self {
        let inner = DomNodeInner {
            node_type: NodeType::Root,
            constraint: Constraint::Min(0),
            children: vec![],
            parent: None,
        };
        let key = DOM_NODES.with(|n| n.borrow_mut().insert(inner));
        Self {
            // inner: Rc::new(RefCell::new(inner)),
            key,
            // parent: Rc::new(RefCell::new(None)),
        }
    }

    pub fn from_fragment(fragment: DocumentFragment) -> Self {
        let inner = DomNodeInner {
            node_type: fragment.node_type,
            constraint: fragment.constraint,
            children: fragment.children,
            parent: None,
        };
        let key = DOM_NODES.with(|n| n.borrow_mut().insert(inner));
        Self {
            // inner:Rc::new(RefCell::new(inner)),
            key,
            // parent: Rc::new(RefCell::new(None)),
        }
    }

    pub fn append_child(&self, node: &DomNode) {
        // *(*node.parent).borrow_mut() = Some(self.key);
        DOM_NODES.with(|d| {
            let mut d = d.borrow_mut();
            d[node.key].parent = Some(self.key);
            d[self.key].children.push(node.key);
        });
        // DOM_NODES.with(|d| d.borrow_mut()[self.key].children.push(node.key));
        // (*self.inner).borrow_mut().children.push(node.key);
    }

    pub fn before(&self, node: &DomNode) {
        // if let Some(parent_id) = self.parent.borrow().as_ref() {
        DOM_NODES.with(|d| {
            let mut d = d.borrow_mut();
            if let Some(parent_id) = d[self.key].parent {
                let parent = d.get_mut(parent_id).unwrap();
                let self_index = parent.children.iter().position(|c| c == &self.key).unwrap();
                parent.children.insert(self_index, node.key);
                d[node.key].parent = Some(parent_id);
                // node.parent = self.parent.clone();
            }
        });
        // }
    }

    pub fn render(&self, frame: &mut Frame, rect: Rect) {
        DOM_NODES.with(|d| d.borrow()[self.key].render(frame, rect));
        // self.inner.borrow().render(frame, rect);
    }
}

impl IntoView for DomNode {
    fn into_view(self) -> View {
        View::DomNode(self)
    }
}

// impl<F> IntoView for DomWidget<F>
// where
//     F: FnMut(&mut Frame, Rect) + 'static,
// {
//     fn into_view(self) -> View {
//         View::new(self.inner)
//     }
// }

// impl IntoView for View {
//     fn into_view(self) -> View {
//         self
//     }
// }

// impl IntoView for DomNode {
//     fn into_view(self) -> View {
//         View::new(move |frame, rect| self.render(frame, rect))
//     }
// }

// pub trait IntoDomNode {
//     fn attach(self, children: &mut Vec<DomNode>);
// }

// impl IntoDomNode for DomNode {
//     fn attach(self, children: &mut Vec<DomNode>) {
//         children.push(self);
//     }
// }

// impl IntoDomNode for View {
//     fn attach(self, children: &mut Vec<DomNode>) {
//         children.push(DomNode::component(self));
//     }
// }

// impl IntoDomNode for Vec<DomNode> {
//     fn attach(self, children: &mut Vec<DomNode>) {
//         for node in self.into_iter() {
//             children.push(node);
//         }
//     }
// }

// impl<F: 'static> IntoDomNode for DomWidget<F>
// where
//     F: FnMut(&mut Frame, Rect),
// {
//     fn attach(self, children: &mut Vec<DomNode>) {
//         self.into_view().attach(children);
//     }
// }

// impl DomNode {
//     fn root() -> Self {
//         Self {
//             id: None,
//             node_type: NodeType::Root,
//             children: vec![],
//             constraint: Constraint::Min(0),
//         }
//     }

//     pub fn component(v: View) -> Self {
//         Self {
//             id: None,
//             node_type: NodeType::Component(RefCell::new(v.into_view())),
//             children: vec![],
//             constraint: Constraint::Min(0),
//         }
//     }

//     pub fn row() -> Self {
//         Self {
//             id: None,
//             node_type: NodeType::Layout {
//                 direction: Direction::Horizontal,
//                 margin: 0,
//             },
//             children: vec![],
//             constraint: Constraint::Min(0),
//         }
//     }

//     pub fn col() -> Self {
//         Self {
//             id: None,
//             node_type: NodeType::Layout {
//                 direction: Direction::Vertical,
//                 margin: 0,
//             },
//             children: vec![],
//             constraint: Constraint::Min(0),
//         }
//     }

//     pub fn overlay() -> Self {
//         Self {
//             id: None,
//             node_type: NodeType::Overlay,
//             children: vec![],
//             constraint: Constraint::Min(0),
//         }
//     }

//     pub fn constraint(mut self, constraint: Constraint) -> Self {
//         self.constraint = constraint;
//         self
//     }

//     pub fn render(&self, frame: &mut Frame, rect: Rect) {
//         match &self.node_type {
//             NodeType::Layout { direction, margin } => {
//                 let layout = Layout::default().direction(*direction).margin(*margin);
//                 let constraints = self.children.iter().map(|c| c.constraint);
//                 let chunks = layout
//                     .constraints(constraints.collect::<Vec<_>>())
//                     .split(rect);
//                 self.children
//                     .iter()
//                     .zip(chunks.iter())
//                     .for_each(|(child, chunk)| {
//                         child.render(frame, *chunk);
//                     });
//             }
//             NodeType::Overlay | NodeType::Root => {
//                 self.children.iter().for_each(|child| {
//                     child.render(frame, rect);
//                 });
//             }
//             NodeType::Component(component) => {
//                 component.borrow().render(frame, rect);
//             }
//         }
//     }

//     pub fn margin(mut self, new_margin: u16) -> Self {
//         if let NodeType::Layout { margin, .. } = &mut self.node_type {
//             *margin = new_margin;
//         }
//         self
//     }

//     pub fn child(mut self, node: impl IntoDomNode) -> Self {
//         node.attach(&mut self.children);
//         self
//     }

//     fn matches_id(&self, id: u32) -> bool {
//         if let NodeType::Component(component) = &self.node_type {
//             return component.borrow().id == id;
//         }
//         false
//     }

//     fn replace_view(&self, view: View) {
//         if let NodeType::Component(component) = &self.node_type {
//             println!("HI");
//             *component.borrow_mut() = view;
//         }
//     }

//     pub fn replace(&self, id: u32, mut new: View) -> Option<View> {
//         if self.matches_id(id) {
//             self.replace_view(new);
//             return None;
//         }
//         for child in self.children.iter() {
//             match child.replace(id, new) {
//                 Some(returned) => {
//                     new = returned;
//                 }
//                 None => {
//                     return None;
//                 }
//             }
//         }
//         Some(new)
//     }
// }

pub fn mount(v: impl IntoView) {
    let node = v.into_view().get_mountable_node();
    DOM.with(|d| d.borrow_mut().append_child(&node));
}

pub fn render_dom(frame: &mut Frame) {
    DOM.with(|d| d.borrow().render(frame, frame.size()));
}
