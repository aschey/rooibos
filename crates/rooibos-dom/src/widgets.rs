use std::sync::atomic::{AtomicU32, Ordering};

use ratatui::prelude::Rect;
use ratatui::style::Style;
use ratatui::text::{Line, Span, Text};
use ratatui::widgets::canvas::{Canvas, Context};
use ratatui::widgets::{StatefulWidget, *};
use ratatui::Frame;
use rooibos_dom_macros::{impl_stateful_render, impl_stateful_widget, impl_widget, make_builder};

use crate::{DomWidget, IntoView};

pub static __NODE_ID: AtomicU32 = AtomicU32::new(1);

#[make_builder]
pub trait MakeBuilder {}

impl_stateful_render!(StatefulRender, visibility=pub);

pub trait NewExt<'a, T>
where
    Self: 'a,
{
    fn new(source: T) -> Self;
}

pub trait NewFrom {}

impl<'a, S, T> NewExt<'a, T> for S
where
    S: NewFrom + 'a,
    Self: From<T>,
{
    fn new(source: T) -> Self {
        Self::from(source)
    }
}

impl<'a> NewFrom for Line<'a> {}
impl<'a> NewFrom for Span<'a> {}
impl<'a> NewFrom for Cell<'a> {}
impl<'a> NewFrom for Text<'a> {}

pub trait StyleExt<'a> {
    fn style(self, style: Style) -> Self;
}

impl<'a> StyleExt<'a> for Span<'a> {
    fn style(mut self, style: Style) -> Self {
        self.style = style;
        self
    }
}

impl<'a> StyleExt<'a> for Text<'a> {
    fn style(self, style: Style) -> Self {
        self.reset_style().patch_style(style)
    }
}

impl<'a> MakeBuilder for Row<'a> {}
impl<'a> MakeBuilder for Cell<'a> {}
impl<'a> MakeBuilder for Span<'a> {}
impl<'a> MakeBuilder for ListItem<'a> {}
impl<'a> MakeBuilder for Line<'a> {}
impl<'a> MakeBuilder for Text<'a> {}
impl<'a> MakeBuilder for Axis<'a> {}
impl MakeBuilder for Style {}
impl MakeBuilder for ListState {}
impl MakeBuilder for TableState {}
impl MakeBuilder for Wrap {}

impl_widget!(Block, visibility=pub, generics=<'a>, make_builder=MakeBuilder);
impl_widget!(Paragraph, visibility=pub, generics=<'a>, make_builder=MakeBuilder);
impl_widget!(List, visibility=pub,generics=<'a>, make_builder=MakeBuilder);
impl_widget!(Tabs, visibility=pub, generics=<'a>, make_builder=MakeBuilder);
impl_widget!(Table, visibility=pub, generics=<'a>, make_builder=MakeBuilder);
impl_widget!(Gauge, visibility=pub, generics=<'a>, make_builder=MakeBuilder);
impl_widget!(LineGauge, visibility=pub, generics=<'a>, make_builder=MakeBuilder);
impl_widget!(BarChart, visibility=pub, generics=<'a>, make_builder=MakeBuilder);
impl_widget!(Clear, visibility=pub, make_builder=MakeBuilder);
impl_widget!(Canvas, visibility=pub, generics=<'a, F>, make_builder=MakeBuilder, where_clause=where F: Fn(&mut Context) + Clone + 'static);
impl_stateful_widget!(List, visibility=pub, generics=<'a>, stateful_render=StatefulRender);
impl_stateful_widget!(Table, visibility=pub, generics=<'a>, stateful_render=StatefulRender);
impl_stateful_widget!(Scrollbar, visibility=pub, generics=<'a>, stateful_render=StatefulRender);

pub trait WrapExt {
    fn trim(self, trim: bool) -> Self;
}

impl WrapExt for Wrap {
    fn trim(self, trim: bool) -> Self {
        Self { trim }
    }
}
