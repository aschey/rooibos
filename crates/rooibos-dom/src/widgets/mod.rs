mod button;
mod chart;
mod sparkline;

use std::any::{Any, TypeId};

pub use button::*;
pub use chart::*;
use ratatui::widgets::{List, Paragraph, Tabs};
pub use sparkline::*;
use taffy::Size;

use crate::MeasureNode;

#[derive(Clone, Copy, PartialEq, Eq)]
pub enum Role {
    Button,
    TextInput,
}

pub trait WidgetRole {
    fn widget_role() -> Option<Role>;
}

impl<T> WidgetRole for T
where
    T: Any,
{
    fn widget_role() -> Option<Role> {
        let current_type = TypeId::of::<Self>();

        if current_type == TypeId::of::<Button>() {
            return Some(Role::Button);
        }
        #[cfg(feature = "input")]
        if current_type == TypeId::of::<tui_textarea::TextArea>() {
            return Some(Role::TextInput);
        }
        None
    }
}

impl MeasureNode for Tabs<'_> {
    fn measure(
        &self,
        known_dimensions: taffy::Size<Option<f32>>,
        available_space: taffy::Size<taffy::AvailableSpace>,
        style: &taffy::Style,
    ) -> taffy::Size<f32> {
        Size::zero()
    }
}

impl MeasureNode for List<'_> {
    fn measure(
        &self,
        known_dimensions: Size<Option<f32>>,
        available_space: Size<taffy::AvailableSpace>,
        style: &taffy::Style,
    ) -> Size<f32> {
        Size::zero()
    }
}
