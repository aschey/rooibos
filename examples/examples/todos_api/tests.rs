use std::time::Duration;

use rooibos::keybind::{CommandFilter, CommandHandler};
use rooibos::reactive::KeyCode;
use rooibos::reactive::dom::widgets::Role;
use rooibos::reactive::dom::{DomNodeRepr, root};
use rooibos::runtime::RuntimeSettings;
use rooibos::tester::TestHarness;

use crate::server::run_server;
use crate::{Command, app};

macro_rules! assert_snapshot {
    ($name:literal, $harness:expr) => {
        let buffer = $harness.buffer().await;
        insta::with_settings!({
            snapshot_path => "./snapshots"
        }, {
            insta::assert_debug_snapshot!($name, buffer);
        });
    };
}

#[rooibos::test]
async fn test_todos() {
    let listener = tokio::net::TcpListener::bind("127.0.0.1:9353")
        .await
        .unwrap();

    let mut cmd_handler = CommandHandler::<Command>::new();
    cmd_handler.generate_commands();

    tokio::spawn(run_server(listener));
    let mut harness = TestHarness::new_with_settings(
        RuntimeSettings::default().handle_commands(cmd_handler),
        50,
        15,
    )
    .await;
    harness.mount(|| app(Duration::from_millis(500))).await;

    let root_layout = root();
    // Wait for initial data load
    harness
        .wait_for(async |harness, _| {
            harness
                .find_by_text(&root_layout, "Add a todo")
                .await
                .is_some()
                && harness
                    .find_by_text(&root_layout, "No todos")
                    .await
                    .is_some()
        })
        .await
        .unwrap();
    assert_snapshot!("init", harness);

    // Add todo
    let todo_input = root_layout.find_all_by_role(Role::TextInput)[0].clone();
    todo_input.focus();
    harness.send_text("todo 1");
    harness.send_key(KeyCode::Enter);

    let todos_block = harness.get_by_block_text(&root_layout, "Todos").await;
    harness
        .wait_for(async |harness, _| harness.find_by_text(&todos_block, "todo 1").await.is_some())
        .await
        .unwrap();
    assert_snapshot!("after_create", harness);

    update_todo(&mut harness, &todos_block, "1", "todo 11", false).await;
    assert_snapshot!("keyboard_update", harness);
    wait_for_notification(&mut harness, "Todo updated").await;

    update_todo(&mut harness, &todos_block, "2", "todo 112", true).await;
    assert_snapshot!("click_update", harness);
    wait_for_notification(&mut harness, "Todo updated").await;

    // Delete todo
    harness.get_by_text(&todos_block, "x").await.click();
    harness
        .wait_for(async |harness, _| {
            harness
                .find_by_text(&todos_block, "No todos")
                .await
                .is_some()
        })
        .await
        .unwrap();
    wait_for_notification(&mut harness, "Todo deleted").await;
    assert_snapshot!("delete", harness);

    harness.exit().await;
}

async fn update_todo(
    harness: &mut TestHarness,
    todos_block: &DomNodeRepr,
    send_text: &str,
    new_text: &str,
    use_keyboard: bool,
) {
    // Update todo
    let edit_node = harness.get_by_text(todos_block, "Û±ûÅ").await;
    edit_node.click();
    harness
        .wait_for(async |_, _| todos_block.find_all_by_role(Role::TextInput).len() == 1)
        .await
        .unwrap();
    harness.send_text(send_text);
    harness
        .wait_for(async |harness, _| harness.find_by_text(todos_block, new_text).await.is_some())
        .await
        .unwrap();

    if use_keyboard {
        harness.send_key(KeyCode::Enter);
    } else {
        edit_node.click();
    }

    harness
        .wait_for(async |harness, _| {
            todos_block.find_all_by_role(Role::TextInput).is_empty()
                && harness.find_by_text(todos_block, new_text).await.is_some()
        })
        .await
        .unwrap();
}

async fn wait_for_notification(harness: &mut TestHarness, text: &str) {
    harness
        .wait_for(async |harness, _| !harness.terminal_view().await.contains(text))
        .await
        .unwrap();
}
