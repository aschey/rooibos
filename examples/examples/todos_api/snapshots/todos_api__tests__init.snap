---
source: examples/examples/todos_api/tests.rs
expression: harness.buffer()
---
Buffer {
    area: Rect { x: 0, y: 0, width: 50, height: 15 },
    content: [
        "                                                  ",
        "            ┌───────────────────────────┐╭──────╮ ",
        " Add a Todo │ Add a todo                ││submit│ ",
        "            └───────────────────────────┘╰──────╯ ",
        " ┌Todos─────────────────────────────────────────┐ ",
        " │No todos                                      │ ",
        " │                                              │ ",
        " │                                              │ ",
        " │                                              │ ",
        " │                                              │ ",
        " │                                              │ ",
        " │                                              │ ",
        " └──────────────────────────────────────────────┘ ",
        "                                                  ",
        "                                                  ",
    ],
    styles: [
        x: 0, y: 0, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 1, fg: Blue, bg: Reset, underline: Reset, modifier: NONE,
        x: 41, y: 1, fg: Gray, bg: Reset, underline: Reset, modifier: NONE,
        x: 49, y: 1, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 2, fg: Blue, bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 2, fg: Reset, bg: Reset, underline: Reset, modifier: REVERSED,
        x: 14, y: 2, fg: DarkGray, bg: Reset, underline: Reset, modifier: NONE,
        x: 40, y: 2, fg: Blue, bg: Reset, underline: Reset, modifier: NONE,
        x: 41, y: 2, fg: Gray, bg: Reset, underline: Reset, modifier: NONE,
        x: 42, y: 2, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 48, y: 2, fg: Gray, bg: Reset, underline: Reset, modifier: NONE,
        x: 49, y: 2, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 3, fg: Blue, bg: Reset, underline: Reset, modifier: NONE,
        x: 41, y: 3, fg: Gray, bg: Reset, underline: Reset, modifier: NONE,
        x: 49, y: 3, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 2, y: 5, fg: Gray, bg: Reset, underline: Reset, modifier: NONE,
        x: 10, y: 5, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
    ]
}
